### PR TITLE
resolve main class automatically

### PR DESCRIPTION
**Pull request by ansyral**
Signed-off-by: xuzho <xuzho@microsoft.com>

Microsoft/vscode-java-debug#9

### CODE REVIEW

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+     * resolve main class and project name if not specified.
+     * @param arguments a list of arguments including project files and project name
+     * @return main class and project name
+     * @throws CoreException when there are error when resolving main class.
+     */
+    public String[] resolveMainClass(List<Object> arguments) throws CoreException {
+        List<String> projectNames;
+        if (arguments.size() > 1 && arguments.get(1) != null) {
+            // project name specified
+            projectNames = new ArrayList<String>();
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: ArrayList<> is better

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        for (String f : projectFiles) {
+            SAXReader reader = new SAXReader();
+            Document document;
+            try {
+                document = reader.read(new File(f));
+                Element rootElm = document.getRootElement();
+                projects.add(rootElm.elementText("name"));
+            } catch (DocumentException e) {
+                // TODO Auto-generated catch block
+                e.printStackTrace();
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: add log like logger.error

Comment (andxu) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: see examples:https://github.com/Microsoft/java-debug/pull/68

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+        projects.add(null);
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: while add null here

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        return projects;
+    }
+
+    private String resolveMainClassCore(String projectName) throws CoreException {
+        IJavaSearchScope searchScope = createSearchScope(projectName);
+        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_CASE_SENSITIVE | SearchPattern.R_EXACT_MATCH);
+        ArrayList<String> uris = new ArrayList<String>();
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: List<String> uris = new ArrayList<>();

Comment (andxu) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: `List<String> uris = new ArrayList<>()`

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java (hunk 2 above)
Comment (andxu) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: document is not closed?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java (hunk 3 above)
Comment (ansyral) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: fall back to workspace scope

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java (hunk 2 above)
Comment (ansyral) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: i don't see any close/dispose method for Reader/document

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                return res;
+            }
+        }
+        return res;
+    }
+
+    private List<String> resolveProjectName(List<String> projectFiles) {
+        List<String> projects = new ArrayList<>();
+        for (String f : projectFiles) {
+            SAXReader reader = new SAXReader();
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: Does eclipse have any built-in utility to help parse .project file?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                    } catch (JavaModelException e) {
+                        // ignore
+                    }
+                }
+            }
+        };
+        SearchEngine searchEngine = new SearchEngine();
+        searchEngine.search(pattern, new SearchParticipant[] {SearchEngine.getDefaultSearchParticipant()},
+                searchScope, requestor, null /* progress monitor */);
+        return uris.size() == 0 ? null : uris.get(0);
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: if there are multiple main class in the target project, should pop up a dropdown menu in UI to tell user to choose one.

Comment (andxu) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: I think we should return multiple to let js side has the possibilities to choose.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java (hunk 5 above)
Comment (ansyral) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: seems not

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 7]
```diff
@@ -111,21 +76,33 @@ public void acceptSearchMatch(SearchMatch match) {
...
-            return SearchEngine.createWorkspaceScope();
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o instanceof ResolutionItem) {
+                ResolutionItem item = (ResolutionItem) o;
+                return item.mainClass == this.mainClass && item.projectName == this.projectName;
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: in java == is not for equal test.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 8]
```diff
@@ -100,7 +57,15 @@ public void acceptSearchMatch(SearchMatch match) {
...
                     try {
                         if (method.isMainMethod()) {
-                            uris.add(method.getDeclaringType().getFullyQualifiedName());
+                            IResource resource = method.getResource();
+                            if (resource != null) {
+                                IProject project = resource.getProject();
+                                if (project != null) {
+                                    String mainClass = method.getDeclaringType().getFullyQualifiedName();
+                                    String projectName = ProjectsManager.DEFAULT_PROJECT_NAME.equals(project.getName()) ? null : project.getName();
+                                    uris.add(new ResolutionItem(mainClass, projectName));
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: rename a reasonable name.

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 9]
```diff
@@ -36,62 +28,27 @@
...
 public class ResolveMainClassHandler {
     private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
 
     /**
-     * resolve main class and project name if not specified.
-     * @param arguments a list of arguments including project files and project name
-     * @return main class and project name
+     * resolve main class and project name.
+     * @return an array of main class and project name
      * @throws CoreException when there are error when resolving main class.
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: errors

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            if (o instanceof ResolutionItem) {
+                ResolutionItem item = (ResolutionItem) o;
+                return item.mainClass == this.mainClass && item.projectName == this.projectName;
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return mainClass.hashCode() * 13 + (projectName == null ? 0 : projectName.hashCode());
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: missing the handle case for null mainClass and null projectName

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    public Object resolveMainClass() throws CoreException {
+        return resolveMainClassCore();
+    }
+
+    private List<ResolutionItem> resolveMainClassCore() throws CoreException {
+        IJavaSearchScope searchScope = SearchEngine.createWorkspaceScope();
+        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_CASE_SENSITIVE | SearchPattern.R_EXACT_MATCH);
+        ArrayList<ResolutionItem> uris = new ArrayList<>();
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: `uris` -> `results`

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java (hunk 7 above)
Comment (ansyral) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: nice catch, i should use equals instead

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * resolve main class and project name.
+     * @return an array of main class and project name
+     * @throws CoreException when there are errors when resolving main class.
+     */
+    public Object resolveMainClass() throws CoreException {
+        return resolveMainClassCore();
+    }
+
+    private List<ResolutionItem> resolveMainClassCore() throws CoreException {
```
Comment (testforstephen) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveMainClassHandler.java]: new method resolveMainClassCore is unnecessary.

**Pull request merged**